### ISSUE TITLE
"incompatible Harmony patch detected"

### ISSUE THREAD

**Issue by maxiskim** [SETNAHQ/Shicho#28]
The game runs slower as time passes, until the simulation basically stops (still runs, but extremely slowly).

According to the output log, there are multiple and continuous incidents of the following error lines:

```
incompatible Harmony patch detected; this might cause serious conflicts (or visual glitch) in-game
Error: System.InvalidCastException: Cannot cast from source type to destination type.
  at Harmony.PatchInfoSerialization.Deserialize (System.Byte[] bytes) [0x00000] in <filename unknown>:0 
  at Harmony.HarmonySharedState.GetPatchInfo (System.Reflection.MethodBase method) [0x00000] in <filename unknown>:0 
  at Harmony.PatchProcessor.GetPatchInfo (System.Reflection.MethodBase method) [0x00000] in <filename unknown>:0 
  at Harmony.HarmonyInstance.GetPatchInfo (System.Reflection.MethodBase method) [0x00000] in <filename unknown>:0 
  at Shicho.Patcher.Util.UnpatchTarget (Shicho.Patcher.PatchPredicate pred) [0x00000] in <filename unknown>:0 

(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
```
I would love to post the entire output log, unfortunately the log is over 150MB(!!).  Other noticable information in the output logs are as follows:

-----------------------------------------------------------------------------

**TLM** 
```
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

GameSettings: 'TransportsLinesManager5_' is not found or cannot be loaded. Make sure to call GameSettings.AddSettingsFile() to register a settings file before using SavedValue().  [Settings - Internal]
```
I believe this refers to an harmless error from TLM mod, probably irrelvant to Shicho.

-----------------------------------------------------------------------------

**Metro Overhaul Mod**

```
NullReferenceException: Object reference not set to an instance of an object
  at MetroOverhaul.UI.MetroStationCustomizerUI.RestoreStationTrackStyles (.BuildingInfo info) [0x00000] in <filename unknown>:0 
  at MetroOverhaul.UI.MetroStationCustomizerUI.Update () [0x00000] in <filename unknown>:0 
UnityEngine.DebugLogHandler:Internal_LogException(Exception, Object)
UnityEngine.DebugLogHandler:LogException(Exception, Object)
UnityEngine.Logger:LogException(Exception, Object)
UnityEngine.Debug:LogException(Exception)
MetroOverhaul.UI.MetroStationCustomizerUI:Update()
 
(Filename:  Line: -1)

```
Contrary to the output log, the NullRef message does not pop up in the game.  I have notified the devs of MetroOverhaul mod for the errors.

Thank you for your wonderful work!

**Comment by saki7** [comment]
This is due to incompatible external mods; at least I need the full list of your enabled mods.

I can't guarantee to fix this kind of error since I'm having no performance issues in my game even after few dozens of hours experimenting. This is a non-reproducible thing, and I can't do better thing than just randomly guessing the mod which is causing this issue.

As you can see, the Harmony unpatch is executed periodically in Shicho so that the game's getting slower as time passes. Although I do have experienced this kind of slowness, however I'm afraid that it's caused by the dumb debugger for the game itself.

Thanks for the report.

**Comment by saki7** [comment]
If anyone can reduce the specific cause of this issue, I'm willing to commit a fix.

**Comment by ntoff** [comment]
I'm also having this issue when Shicho and [Klyte Commons](https://steamcommunity.com/sharedfiles/filedetails/?id=1312735149) are both enabled. I unsubscribed from my 70ish  mods and went through them a few at a time, narrowed it down to klyte commons. 

Subscribed to mods: 
* Klyte Commons
* Loading screen mod [test]
* Shicho

and the vanilla mods for unlimited money and unlock everything are enabled.

Then I just started a new game and I get a massive log full of the incompatible patch error. 

Hope this somehow helps you find a fix (or just list that mod as being incompatible).

**Comment by saki7** [comment]
So it proves that my assumption for the error was correct.

> ![image](https://user-images.githubusercontent.com/1901813/49391572-61db6980-f76f-11e8-93f3-9cb1f3bfc0ec.png)
> _From the comment of Klyte Commons mod_

You could see that some other person is complaining about the Harmony version incompatibility. It indicates that we (Shicho and other incompatible mods) are in the same situation; we could do nothing about it. Sorry for the inconvenience.

Note: I *could* remove the "massive" error logs in Shicho for sure, however it means that you'll get a non-working Shicho mod in your game which is totally useless, so it makes no difference than just unsubscribing from Shicho.

**Comment by saki7** [comment]
The only solution for this issue is to ask the author of incompatible mods to update their Harmony dependency.

Shicho had been using Harmony's recent master branch (which is newer than v1.2.0.1). Four months have passed since the last Harmony release, and we can't just rollback our dependency to older ones.

**Comment by ntoff** [comment]
>you'll get a non-working Shicho mod in your game which is totally useless,

It actually does mostly still work, some options I'm not entirely sure what they do so I can't tell if they're working or not. The only reason I discovered the error in the log was just by pure chance that I was *extremely* bored and decided to look at the log.

> so it makes no difference than just unsubscribing from Shicho.

I unsubbed from klyte commons and "touch this" instead, I'll never give up Shicho, I love it too much.
